File: neetcode/blind75/trees/subtree_of_a_binary_tree.py
# https://neetcode.io/problems/subtree-of-a-binary-tree

# A subtree of a binary tree tree is a tree that consists of a node in tree and all of this node's descendants. The tree tree could also be considered as a subtree of itself.

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:   
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:

        if (not root):
            return False
        if (not subRoot):
            return True

        if (self.checkSubTree(root, subRoot)):
            # if subtree exists
            return True

        # try with different node as the root
        return (self.isSubtree(root.left, subRoot) or self.isSubtree(root.right, subRoot))

    def checkSubTree(self, root, subRoot):
        if (not root and not subRoot):
            return True
        elif (root and subRoot and root.val == subRoot.val):
            return self.checkSubTree(root.left, subRoot.left) and self.checkSubTree(root.right, subRoot.right)
        else:
            return False

        

File: neetcode/neet150_JS/Trees/subtree-of-a-binary-tree.js
// https://neetcode.io/problems/subtree-of-a-binary-tree

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {

    checkSubTree(root, subRoot) {
        if (root === null && subRoot === null) {
            return true
        }

        if (root !== null && subRoot !== null && root.val === subRoot.val) {
            return (this.checkSubTree(root.left, subRoot.left) && this.checkSubTree(root.right, subRoot.right))
        } else {
            return false
        }
    }

    DFS(root, subRoot) {
        if (root === null) {
            return false
        }

        if (this.checkSubTree(root, subRoot)) {
            return true
        }

        return (this.DFS(root.left, subRoot) || this.DFS(root.right, subRoot))
    }

    /**
     * @param {TreeNode} root
     * @param {TreeNode} subRoot
     * @return {boolean}
     */
    isSubtree(root, subRoot) {
        if (root === null) {
            return false
        }
        if (subRoot === null) {
            return true
        }

        return this.DFS(root, subRoot)
    }
}
